File: avalon-modules/avalon-dataset/avalon-dataset-repository/src/test/java/io/github/lishangbu/avalon/dataset/repository/BerryFlavorRepositoryTest.java
package io.github.lishangbu.avalon.dataset.repository;

import io.github.lishangbu.avalon.dataset.entity.BerryFlavor;
import jakarta.annotation.Resource;
import java.util.List;
import java.util.Optional;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.data.domain.Example;

/// 树果风味数据访问层测试
class BerryFlavorRepositoryTest extends AbstractRepositoryTest {

    @Resource private BerryFlavorRepository berryFlavorRepository;

    /// 验证 通过ID查询
    @Test
    void shouldFindBerryFlavorById() {
        // Act
        Optional<BerryFlavor> flavorOptional = berryFlavorRepository.findById(1L);
        Assertions.assertTrue(flavorOptional.isPresent());
        BerryFlavor flavor = flavorOptional.get();
        // Assert
        Assertions.assertNotNull(flavor, "应已插入 id=1 的 属性为辣的树果风味记录");
        Assertions.assertEquals(1L, flavor.getId());
        Assertions.assertEquals("spicy", flavor.getInternalName());
        Assertions.assertEquals("辣", flavor.getName());
    }

    /// 验证 selectList 的动态查询能够基于部分 internalName 匹配到 数据库中的数据
    @Test
    void shouldSelectListMatchByInternalName() {
        // Arrange - 构造查询条件，使用部分 internalName
        BerryFlavor cond = new BerryFlavor();
        cond.setInternalName("spicy");

        // Act
        List<BerryFlavor> results = berryFlavorRepository.findAll(Example.of(cond));

        // Assert
        Assertions.assertNotNull(results);
        Assertions.assertFalse(results.isEmpty(), "应至少匹配到一条属性为辣的树果风味记录");
        Assertions.assertTrue(results.stream().anyMatch(r -> "spicy".equals(r.getInternalName())));
    }

    /// 验证 selectList 在空条件下返回所有预加载的记录
    @Test
    void shouldReturnAllPreloadedBerryFlavorsWhenNoCondition() {
        // Act
        List<BerryFlavor> all = berryFlavorRepository.findAll();

        // Assert
        Assertions.assertNotNull(all);
        // 根据 db/changelog 数据文件，至少有 5 条记录
        Assertions.assertTrue(all.size() >= 5, "预期至少有 5 个预加载的树果风味记录");
    }

    /// 测试插入树果风味（使用简短合理的 internalName 避免与预加载数据冲突）
    @Test
    void shouldInsertBerryFlavor() {
        // Arrange
        BerryFlavor bf = new BerryFlavor();
        bf.setInternalName("savory");
        bf.setName("鲜美");

        // Act
        berryFlavorRepository.save(bf);

        // Assert
        Assertions.assertNotNull(bf.getId());
        Optional<BerryFlavor> berryFlavorOptional = berryFlavorRepository.findById(bf.getId());
        Assertions.assertTrue(berryFlavorOptional.isPresent());
        BerryFlavor berryFlavor = berryFlavorOptional.get();
        Assertions.assertNotNull(berryFlavor);
        Assertions.assertEquals("savory", berryFlavor.getInternalName());
        Assertions.assertEquals("鲜美", berryFlavor.getName());
    }

    /// 测试根据 ID 更新树果风味的名称
    @Test
    void shouldUpdateBerryFlavorById() {
        // Arrange - 先插入一条临时记录
        BerryFlavor bf = new BerryFlavor();
        bf.setInternalName("fishy");
        bf.setName("鱼腥");
        berryFlavorRepository.saveAndFlush(bf);
        Long id = bf.getId();

        // 修改名称
        bf.setName("鱼腥味");

        // Act
        berryFlavorRepository.saveAndFlush(bf);

        // Assert
        Optional<BerryFlavor> updatedEntityOptional = berryFlavorRepository.findById(id);
        Assertions.assertTrue(updatedEntityOptional.isPresent());
        BerryFlavor updatedEntity = updatedEntityOptional.get();
        Assertions.assertNotNull(updatedEntity);
        Assertions.assertEquals("鱼腥味", updatedEntity.getName());
    }

    /// 测试根据 ID 删除树果风味
    @Test
    void shouldDeleteBerryFlavorById() {
        // Arrange - 插入临时记录
        BerryFlavor bf = new BerryFlavor();
        bf.setInternalName("pungent");
        bf.setName("刺激性风味");
        berryFlavorRepository.saveAndFlush(bf);
        Long id = bf.getId();

        // Act
        berryFlavorRepository.deleteById(id);
        berryFlavorRepository.flush();
        // 验证通过 internalName 查询不到该记录
        BerryFlavor cond = new BerryFlavor();
        cond.setInternalName("pungent");
        List<BerryFlavor> results = berryFlavorRepository.findAll(Example.of(cond));
        Assertions.assertTrue(results.isEmpty(), "删除后按 internalName 查询应返回空集合");
    }
}

File: avalon-modules/avalon-dataset/avalon-dataset-repository/src/test/kotlin/io/github/lishangbu/avalon/dataset/repository/BerryFlavorRepositoryTest.kt
package io.github.lishangbu.avalon.dataset.repository

import io.github.lishangbu.avalon.dataset.entity.BerryFlavor
import io.github.lishangbu.avalon.dataset.entity.dto.BerryFlavorSpecification
import jakarta.annotation.Resource
import org.babyfish.jimmer.sql.ast.mutation.SaveMode
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

/** 树果风味仓储测试 */
class BerryFlavorRepositoryTest : AbstractRepositoryTest() {
    @Resource
    private lateinit var berryFlavorRepository: BerryFlavorRepository

    /** 验证可按 ID 查询树果风味 */
    @Test
    fun shouldFindBerryFlavorById() {
        // Act
        val flavor = requireNotNull(berryFlavorRepository.loadViewById(1L))
        // Assert
        Assertions.assertEquals("1", flavor.id)
        Assertions.assertEquals("spicy", flavor.internalName)
        Assertions.assertEquals("辣", flavor.name)
    }

    /** 验证可按 internalName 查询树果风味 */
    @Test
    fun shouldSelectListMatchByInternalName() {
        // Arrange - 构造查询条件，使用部分 internalName
        val specification = BerryFlavorSpecification(internalName = "spicy")

        // Act
        val results = berryFlavorRepository.listViews(specification)

        // Assert
        Assertions.assertNotNull(results)
        Assertions.assertFalse(results.isEmpty(), "应至少匹配到一条属性为辣的树果风味记录")
        Assertions.assertTrue(results.any { it.internalName == "spicy" })
    }

    /** 验证空条件下返回全部预加载记录 */
    @Test
    fun shouldReturnAllPreloadedBerryFlavorsWhenNoCondition() {
        // Act
        val all = berryFlavorRepository.listViews(null)

        // Assert
        Assertions.assertNotNull(all)
        // 根据 db/changelog 数据文件，至少有 5 条记录
        Assertions.assertTrue(all.size >= 5, "预期至少有 5 个预加载的树果风味记录")
    }

    /** 验证可插入树果风味 */
    @Test
    fun shouldInsertBerryFlavor() {
        // Arrange
        val bf =
            BerryFlavor {
                internalName = "savory"
                name = "鲜美"
            }

        // Act
        val saved = berryFlavorRepository.save(bf, SaveMode.INSERT_ONLY)

        // Assert
        Assertions.assertNotNull(saved.id)
        val berryFlavor = requireNotNull(berryFlavorRepository.loadViewById(saved.id))
        Assertions.assertEquals("savory", berryFlavor.internalName)
        Assertions.assertEquals("鲜美", berryFlavor.name)
    }

    /** 验证可按 ID 更新树果风味名称 */
    @Test
    fun shouldUpdateBerryFlavorById() {
        // Arrange - 先插入一条临时记录
        val created =
            berryFlavorRepository.save(
                BerryFlavor {
                    internalName = "fishy"
                    name = "鱼腥"
                },
                SaveMode.INSERT_ONLY,
            )
        val id = created.id

        // Act
        berryFlavorRepository.save(BerryFlavor(created) { name = "鱼腥味" }, SaveMode.UPSERT)

        // Assert
        val updatedEntity = requireNotNull(berryFlavorRepository.findNullable(id))
        Assertions.assertEquals("鱼腥味", updatedEntity.name)
    }

    /** 验证可按 ID 删除树果风味 */
    @Test
    fun shouldDeleteBerryFlavorById() {
        // Arrange - 插入临时记录
        val created =
            berryFlavorRepository.save(
                BerryFlavor {
                    internalName = "pungent"
                    name = "刺激性风味"
                },
                SaveMode.INSERT_ONLY,
            )
        val id = created.id

        // Act
        berryFlavorRepository.deleteById(id)
        // 验证通过 internalName 查询不到该记录
        val specification = BerryFlavorSpecification(internalName = "pungent")
        val results = berryFlavorRepository.listViews(specification)
        Assertions.assertTrue(results.isEmpty(), "删除后按 internalName 查询应返回空集合")
    }
}
